rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/telemetry/TbMsgTimeseriesNodeTest.java

```diff
@@ -0,0 +1,242 @@
+    private final TenantProfileId TENANT_PROFILE_ID = new TenantProfileId(UUID.fromString("ab78dd78-83d0-43fa-869f-d42ec9ed1744"));
+
+    private TbMsgTimeseriesNode node;
+    private TbMsgTimeseriesNodeConfiguration config;
+    private long tenantProfileDefaultStorageTtl;
+
+    @Mock
+    private TbContext ctxMock;
+    @Mock
+    private RuleEngineTelemetryService telemetryServiceMock;
+
+    @BeforeEach
+    public void setUp() throws TbNodeException {
+        node = new TbMsgTimeseriesNode();
+        config = new TbMsgTimeseriesNodeConfiguration().defaultConfiguration();
+    }
+
+    @ParameterizedTest
+    @EnumSource(TbMsgType.class)
+    void givenMsgTypeAndEmptyMsgData_whenOnMsg_thenVerifyFailureMsg(TbMsgType msgType) throws TbNodeException {
+        init();
+        TbMsg msg = TbMsg.newMsg(msgType, DEVICE_ID, TbMsgMetaData.EMPTY, TbMsg.EMPTY_JSON_ARRAY);
+        node.onMsg(ctxMock, msg);
+        ArgumentCaptor<Throwable> throwableCaptor = ArgumentCaptor.forClass(Throwable.class);
+        verify(ctxMock).tellFailure(eq(msg), throwableCaptor.capture());
+
+        if (TbMsgType.POST_TELEMETRY_REQUEST.equals(msgType)) {
+            assertThat(throwableCaptor.getValue()).isInstanceOf(IllegalArgumentException.class).hasMessage("Msg body is empty: " + msg.getData());
+            return;
+        }
+        assertThat(throwableCaptor.getValue()).isInstanceOf(IllegalArgumentException.class).hasMessage("Unsupported msg type: " + msgType);
+        verifyNoMoreInteractions(ctxMock);
+    }
+
+    @Test
+    void givenTtlFromConfigIsZeroAndUseServiceTsIsTrue_whenOnMsg_thenSaveTimeseriesUsingTenantProfileDefaultTtl() throws TbNodeException {
+        config.setUseServerTs(true);
+        init();
+        String data = """
+                {
+                    "temp": 45,
+                    "humidity": 77
+                }
+                """;
+        TbMsg msg = TbMsg.newMsg(TbMsgType.POST_TELEMETRY_REQUEST, DEVICE_ID, TbMsgMetaData.EMPTY, data);
+
+        when(ctxMock.getTelemetryService()).thenReturn(telemetryServiceMock);
+        when(ctxMock.getTenantId()).thenReturn(TENANT_ID);
+        doAnswer(invocation -> {
+            TelemetryNodeCallback callback = invocation.getArgument(5);
+            callback.onSuccess(null);
+            return null;
+        }).when(telemetryServiceMock).saveAndNotify(any(), any(), any(), anyList(), anyLong(), any());
+
+        node.onMsg(ctxMock, msg);
+
+        ArgumentCaptor<List<TsKvEntry>> entryListCaptor = ArgumentCaptor.forClass(List.class);
+        verify(telemetryServiceMock).saveAndNotify(eq(TENANT_ID), isNull(), eq(DEVICE_ID), entryListCaptor.capture(),
+                eq(tenantProfileDefaultStorageTtl), any(TelemetryNodeCallback.class));
+        List<TsKvEntry> entryListCaptorValue = entryListCaptor.getValue();
+        assertThat(entryListCaptorValue.size()).isEqualTo(2);
+        verifyTimeseriesToSave(entryListCaptorValue, msg);
```

Comment:
This logic can be replaced with lists comparison. You can build a list in the same way as rule node does this. See how we build List<TsKvEntry> in the rule node.

The only one field that will be different is a ts field for each TsKvEntry and here is where power of AssertJ library can be used. See method: usingRecursiveFieldByFieldElementComparatorIgnoringFields and containsExactlyElementsOf